### ISSUE TITLE
Adding a new section broke xfixup targets.

### ISSUE THREAD

**Issue by paulhuggett** [SNSystems/llvm-project-prepo#73]
It seems as though llvm-project-prepo commit 78b25f5ea123e4f8d1bc655592360444b1200b21 (which pairs with pstore [`ccf8c68`](https://github.com/SNSystems/pstore/commit/ccf8c685f4e5dcc022cee1c7023c7ae8fed27b17) and rld [`a57fb6d`](https://github.com/SNSystems/rld/commit/a57fb6dec7eb2d319e62850217149d9bb82ac77a)) that added a new `interp` section to the repository’s fragments has broken `repo2obj` in some way. The only known symptom right now is that an attempt to build LLVM with the prepo compiler results in a build of `clang-tblgen` which crashes.

**Comment by paulhuggett** [comment]
I wrote a shell script ([here](https://gist.github.com/paulhuggett/39b92f1dd5123bedc479d033b13fdcde)) to do four jobs:
1. Build the “latest” compiler from the HEAD of the pstore and llvm-project-prepo trees. This includes the changes listed above that are under suspicion.
1. Build a “good” compiler from the commit prior to suspicious change.
1. Use the “good” compiler to build LLVM.
1. Use the “latest” compiler to build LLVM.

We anticipate that the last step will fail.

**Comment by paulhuggett** [comment]
Contining to investigate. It seems that a “good” compiler at commits before the ones listed [above](#issue-565913913) still fails to build LLVM although the symptoms seem to be somewhat unpredictable. 

I’ve wound the repository back as far as October 30th (that is, https://github.com/SNSystems/llvm-project-prepo/commit/7c3c8e8a0819120b06539e93159b454ce755e2e8 and  https://github.com/SNSystems/pstore/commit/88ed1f2aa43d4340a95fede5df49b83e7feef55e) and am seeing similar problems. Most recently:

~~~
[1802/2968] Linking CXX static library lib/libclangDynamicASTMatchers.a
/usr/bin/ar: tools/clang/lib/ASTMatchers/Dynamic/CMakeFiles/clangDynamicASTMatchers.dir/Registry.cpp.o.elf: invalid entry in SHT_GROUP section [1658]
/usr/bin/ar: tools/clang/lib/ASTMatchers/Dynamic/CMakeFiles/clangDynamicASTMatchers.dir/Registry.cpp.o.elf: invalid entry in SHT_GROUP section [1658]
/usr/bin/ar: tools/clang/lib/ASTMatchers/Dynamic/CMakeFiles/clangDynamicASTMatchers.dir/Registry.cpp.o.elf: SHT_GROUP section [index 1658] has no SHF_GROUP sections
/usr/bin/ar: tools/clang/lib/ASTMatchers/Dynamic/CMakeFiles/clangDynamicASTMatchers.dir/Registry.cpp.o.elf: invalid entry in SHT_GROUP section [1658]
/usr/bin/ar: tools/clang/lib/ASTMatchers/Dynamic/CMakeFiles/clangDynamicASTMatchers.dir/Registry.cpp.o.elf: invalid entry in SHT_GROUP section [1658]
/usr/bin/ar: tools/clang/lib/ASTMatchers/Dynamic/CMakeFiles/clangDynamicASTMatchers.dir/Registry.cpp.o.elf: SHT_GROUP section [index 1658] has no SHF_GROUP sections
/usr/bin/ranlib: lib/libclangDynamicASTMatchers.a(Registry.cpp.o.elf): invalid entry in SHT_GROUP section [1658]
/usr/bin/ranlib: lib/libclangDynamicASTMatchers.a(Registry.cpp.o.elf): invalid entry in SHT_GROUP section [1658]
/usr/bin/ranlib: lib/libclangDynamicASTMatchers.a(Registry.cpp.o.elf): SHT_GROUP section [index 1658] has no SHF_GROUP sections
/usr/bin/ranlib: lib/libclangDynamicASTMatchers.a(Registry.cpp.o.elf): invalid entry in SHT_GROUP section [1658]
/usr/bin/ranlib: lib/libclangDynamicASTMatchers.a(Registry.cpp.o.elf): invalid entry in SHT_GROUP section [1658]
/usr/bin/ranlib: lib/libclangDynamicASTMatchers.a(Registry.cpp.o.elf): SHT_GROUP section [index 1658] has no SHF_GROUP sections
~~~

The symptoms do point to a bug in repo2obj, but a reasonably long-standing one.

**Comment by paulhuggett** [comment]
> The symptoms do point to a bug in repo2obj, but a reasonably long-standing one.

I’ve now reproduced this problem in isolation.

**Comment by paulhuggett** [comment]
> I’ve now reproduced this problem in isolation.

This turns out to be an instance of issue #67. The compilation results in `rep2obj` creating 67195 sections. This number is more than the `Elf_Half` (16 bit) `e_shnum` can hold.

**Comment by paulhuggett** [comment]
Following further discussion, I’ve come to the conclusion that there is more than one problem happening here.
- The build is resulting in ELF files with > SHN_LORESERVE (0xff00) sections. Repo2obj did not support SHN_XINDEX and extended section numbers that are necessary to support this in ELF. That’s the subject of issue #67 and I won’t discuss it further here.
- Changing the order of the section type enum (which was pstore [`ccf8c68`](https://github.com/SNSystems/pstore/commit/ccf8c685f4e5dcc022cee1c7023c7ae8fed27b17)) is causing tblgen to crash. (Note that at the time of writing I haven’t been able to reproduce this crash.)

In the program repository design, an individual “fragment” is intended to represent an individual function or global data object. To do that, the fragment can contain multiple “sections” each of which has a defined type and meaning. Connections between the different sections of a fragment can be efficiently represented as “internal fixups” (or “ifixups”). In contrast, connections between different fragments as represented as “external fixups” (or “xfixups”).

An xfixup is a reference to a fragment by name. A fragment is named by an entry in a compilation which references it, and an individual fragment may have more than one name. This suggests a question: when an xfixup references a fragment, to which section of the fragment does it refer? The answer is that it refers to the _lowest numbered section with content_. This implies that the section numbers play a crucial role in connecting fragments together.

From the above we can see that the change to the section order was a breaking change and the original needs to be restored.

Here’s an example. Start with the source code below:
~~~C
int f (void) {
    static int g = 0;
    return g++;
}
int g (void) {
    return f () + 1;
}
~~~
This is most efficiently represented in the program repository as two fragments:

~~~
| compilation |            
| f           | -----> | fragment |
| g           | --\    | text     | i:data+0
                  |    | data     |
                  |
                  \--> | fragment |
                       | text     | x:"f"+0
~~~
As you can (hopefully) see, we have two fragments: one (”f’) with two sections (text and data) and an ifixup from text to data; and a second (”g”) with a single text section and an xfixup referencing “f”. This latter xfixup is the description of the call from `g()` to `f()` which must clearly point to the text section of this fragment.

**Comment by paulhuggett** [comment]
This should be fixed by pstore [`17569c9`](https://github.com/SNSystems/pstore/commit/17569c922312feb58ede9367f78a673f9679e64c) which restores the original order of names in the `section_type` enum.

**Comment by MaggieYingYi** [comment]
> 
> 
> This should be fixed by pstore [`17569c9`](https://github.com/SNSystems/pstore/commit/17569c922312feb58ede9367f78a673f9679e64c) which restores the original order of names in the `section_type` enum.

I have built upstream LLVM+clang project using the latest commit (17569c9) . The segmentation fault issue (llvm-tblgen) has been fixed.

**Comment by paulhuggett** [comment]
Closing.
